firewood/src/merkle/stream.rs

```diff
@@ -1,290 +1,398 @@
+                        let value = value.to_vec();
+
+                        Ok((std::mem::take(key), value))
+                    });
 
-                self.key_state = IteratorState::Iterating { last_node, parents };
+                    visited_node_path.push((found_node, 0));
 
-                current_key
+                    self.key_state = IteratorState::Iterating { visited_node_path };
+
+                    return Poll::Ready(next_result);
+                }
+
+                self.key_state = IteratorState::Iterating { visited_node_path };
+
+                self.poll_next(_cx)
             }
-            IteratorState::StartAtKey(key) => {
-                // TODO: support finding the next key after K
-                let root_node = self
-                    .merkle
-                    .get_node(self.merkle_root)
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                let (found_node, parents) = self
-                    .merkle
-                    .get_node_and_parents_by_key(root_node, &key)
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+            IteratorState::Iterating { visited_node_path } => {
+                let next = find_next_result(merkle, visited_node_path)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))
+                    .transpose();
 
-                let Some(last_node) = found_node else {
-                    return Poll::Ready(None);
-                };
+                Poll::Ready(next)
+            }
+        }
+    }
+}
+
+enum NodeRef<'a> {
+    New(ObjRef<'a>),
+    Visited(ObjRef<'a>),
+}
+
+#[derive(Debug)]
+enum InnerNode<'a> {
+    New(&'a NodeType),
+    Visited(&'a NodeType),
+}
+
+impl<'a> NodeRef<'a> {
+    fn inner(&self) -> InnerNode<'_> {
+        match self {
+            Self::New(node) => InnerNode::New(node.inner()),
+            Self::Visited(node) => InnerNode::Visited(node.inner()),
+        }
+    }
+
+    fn into_node(self) -> ObjRef<'a> {
+        match self {
+            Self::New(node) => node,
+            Self::Visited(node) => node,
+        }
+    }
+}
+
+fn find_next_result<'a, S: ShaleStore<Node>, T>(
+    merkle: &'a Merkle<S, T>,
+    visited_path: &mut Vec<(ObjRef<'a>, u8)>,
+) -> Result<Option<(Key, Value)>, super::MerkleError> {
+    let next = find_next_node_with_data(merkle, visited_path)?.map(|(next_node, value)| {
+        let partial_path = match next_node.inner() {
+            NodeType::Leaf(leaf) => leaf.path.iter().copied(),
+            NodeType::Extension(extension) => extension.path.iter().copied(),
+            _ => [].iter().copied(),
+        };
+
+        let key = key_from_nibble_iter(nibble_iter_from_parents(visited_path).chain(partial_path));
+
+        visited_path.push((next_node, 0));
+
+        (key, value)
+    });
+
+    Ok(next)
+}
+
+fn find_next_node_with_data<'a, S: ShaleStore<Node>, T>(
```

Comment:
nit: as the core function, I feel it deserves some high level comments describe how things work.